File: 20. 有效的括号/main.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  let stack = [];
  let map = {
    ")": "(",
    "]": "[",
    "}": "{",
  };
  for (let i = 0; i < s.length; i++) {
    if (map[s[i]]) {
      // 如果是右括号，则出栈
      let data = stack.pop();
      if (data != map[s[i]]) return false; // 如果出栈的元素不等于对应的左括号，则返回 false
    } else {
      stack.push(s[i]); // 如果是左括号，则入栈
    }
  }
  return stack.length === 0;
};

File: 20. 有效的括号/main.rs
use std::collections::HashMap;
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut st = Vec::with_capacity(s.len());
        // let map = HashMap::from([
        //     (')', '('),
        //     ('}', '{'),
        //     (']', '['),
        // ]);
        let map: HashMap<char, char> = [
            (')', '('),
            (']', '['),
            ('}', '{'),
        ]
            .into_iter()
            .collect();
        for c in s.chars() {
            if let Some(&need) = map.get(&c) {
                // 如果栈为空，说明没有匹配的左括号
                if let Some(top) = st.pop() {
                    if top != need {
                        return false;
                    }
                } else {
                    return false;
                }
            } else if c == '(' || c == '{' || c == '[' {
                st.push(c);
            } else {
                return false;
            }
        }
        st.is_empty()
    }
}
